### ISSUE TITLE
Plan to publish app to F-Droid official repo

### ISSUE THREAD

**Issue by papjul** [breezy-weather/breezy-weather#702]
**Update:** Solution was redesigned compared to my initial message to ensure we can get Breezy Weather to F-Droid default repository as soon as possible, following the removal of Geometric Weather from F-Droid default repository.

This is the current checklist:

* [X] Release 5.0-alpha

* [x] ~~Wait for v5.0 to reach stable state~~ Fix #770 

* [X] Create a `fdroid` flavor that is not able to connect to non-free (weather) sources. This means only keeping Open-Meteo and (once it becomes open-source) Pirate Weather

* [x] Make a custom Fastlane for F-Droid default repository

* [x] Complete all issues from [milestone](https://github.com/breezy-weather/breezy-weather/milestone/2)

* [x] Remove non-deterministic stuff to allow for reproducible builds

* [ ] Reviewing internally the release strategy

* [ ] Wait for F-Droid team to release our APK

**Comment by papjul** [comment]
*Reserved*

**Comment by licaon-kter** [comment]
@papjul 
> it contains two proprietary dependencies and proprietary fonts, so it might get removed from F-Droid at some point.

links please

**Comment by papjul** [comment]
org.greenrobot:greendao-gradle-plugin - greenrobot/greenDAO#412
And I thought com.github.didikee:AndroidDonate was as well as I couldn't find the sources and remember replacing it, but now that I look again, I seem to find the sources in https://github.com/didikee/AndroidDonate/tree/master/donate

If you intend to remove Geometric Weather, then I think I will just end up making a build variant with no non-free networks (all weather sources except Open-Meteo).
I think it’s a good compromise, as:
- Weather sources are the elements that can break the more easily, and you know I don’t like how much you have to wait in F-Droid main repository to get updates. It tarnishes the reputation of the app by making users think we can’t fix critical issues fast.
- Open-Meteo is a rather well-tested and reliable weather source, so external factors breaking the app are unlikely
- This version will only connect to free networks which make it even more FOSS (even though non-free networks are not forbidden and you don’t always tag such apps as “non-free networks”)

And I will replace the logo with a free basic one, although F-Droid usually don’t tag apps as non-free assets if the logo is non-free (as seen in MPL-licensed projects).

To achieve that:
- New build variant
- Create a separate GitHub repo so we can have a different fastlane
- Which will also allow to push new versions at different times. For example: make users of the standard version get the major upgrades first so they can test it well before we push it to the F-Droid version (which should always get the latest most stable release)

**Comment by licaon-kter** [comment]
@papjul thanks for the info, disabled the app

**Comment by IzzySoft** [comment]
@papjul

> Create a separate GitHub repo so we can have a different fastlane

Fastlane supports build variants, see my [Fastlane Cheat Sheet](https://gitlab.com/-/snippets/1895688), especially the [section on locations](https://gitlab.com/-/snippets/1895688#locations). `/src/<buildFlavor>/fastlane/metadata/android/<locale>/` would fit here then for F-Droid IMHO.

> allow to push new versions at different times

Could also be achieved by using special tag name "prefixes" – e.g. have F-Droid bound to tags starting with `fdroid-` and ignoring all others:

```yaml
UpdateCheckMode: Tags ^fdroid-.*$
```

Just giving you the options – of course your decision, both ways should work.

**Comment by papjul** [comment]
Thank you @IzzySoft, it will definitely be useful!

**Comment by papjul** [comment]
> Could also be achieved by using special tag name "prefixes" – e.g. have F-Droid bound to tags starting with `fdroid-` and ignoring all others:
> 
> ```yaml
> UpdateCheckMode: Tags ^fdroid-.*$
> ```

Is the following regexp valid for you?

```yaml
UpdateCheckMode: Tags ^.*[0-9]$
```

This excludes anything that doesn’t end with a number,excluding any version ending with `-beta` or `-alpha` for example.

**Comment by licaon-kter** [comment]
if you never have `-alpha7` it might be fine :smile:

**Comment by IzzySoft** [comment]
> if you never have -alpha7 it might be fine 😄

Why? If @papjul just wants to exclude the unnumbered ones but numbered alphas/betas are fine to ship… :speak_no_evil: :dash:

**Comment by licaon-kter** [comment]
I tested the regex with an online checker, it picked up `1.0-alpha7` too

Maybe I did it wrong?

**Comment by papjul** [comment]
It's correct. I only put numbers first. The version scheme is the following:
- Name: v5.1.3 or v5.1.3-alpha or v5.1.3-beta
- Version code: 50103 (whether it's alpha or not)

If I need to bump version, it becomes:
- Name: v5.1.4 or v5.1.4-alpha or v5.1.4-beta
- Version code: 50104 (whether it's alpha or not)

Having numbers after the `-alpha` or `-beta` suffix would be confusing with the way I handle version code. Of course, if one day, I have v5.99.99 (59999), I will have to bump to v6.0.0 next time or I will run out of version codes for v5.

**Comment by IzzySoft** [comment]
```yaml
UpdateCheckMode: Tags ^v[0-9\.]$
```

**Comment by papjul** [comment]
Alright, `fdroid` flavor is ready, will request for packaging once v5 feels stable enough.

> Fastlane supports build variants, see my [Fastlane Cheat Sheet](https://gitlab.com/-/snippets/1895688), especially the [section on locations](https://gitlab.com/-/snippets/1895688#locations). `/src/<buildFlavor>/fastlane/metadata/android/<locale>/` would fit here then for F-Droid IMHO.

Alright, now I have both `fastlane/` and `src/fdroid/fastlane/`. Is it OK or should I move `fastlane/` to `src/basic/fastlane/` and you update IzzyOnDroid and I update my F-Droid repo to use the `basic` flavor?

**Comment by IzzySoft** [comment]
Up to you. With my repo I can configure *any* entry-point for Fastlane. I just need to know which I should use. So if I shall switch, just let me know where to.

**Comment by papjul** [comment]
> Up to you. With my repo I can configure _any_ entry-point for Fastlane. I just need to know which I should use. So if I shall switch, just let me know where to.

@IzzySoft For the clarity of the repo, I moved the `fastlane/` repo into `src/basic/fastlane/`. I kept an old `fastlane/` location for the transition. If I remember correctly, you will pull the changes on next GitHub release? The new `src/basic/fastlane/` has updated screenshots to confirm it’s taken into account.

**Comment by IzzySoft** [comment]
So which tree is intended to be used with my repo then? I assume `src/basic/fastlane`, as the other one is named `fdroid`? I've set that now – and yes, it would take effect with the next release (or anytime earlier if you want me to manually trigger a fetch, which can be done here easily).

**Comment by papjul** [comment]
Correct, `basic` is the fastlane for all repos except F-Droid default repository which will use `fdroid` (which should have been called `fdroiddefaultrepository` but a short name was more convenient).

Now that I think of it, you might need to change your regexp starting from next version.

This is the list of assets:
```
breezy-weather-${{ env.VERSION_TAG }}-standard.apk
breezy-weather-arm64-v8a-${{ env.VERSION_TAG }}-standard.apk
breezy-weather-armeabi-v7a-${{ env.VERSION_TAG }}-standard.apk
breezy-weather-x86-${{ env.VERSION_TAG }}-standard.apk
breezy-weather-x86_64-${{ env.VERSION_TAG }}-standard.apk
mapping-${{ env.VERSION_TAG }}-standard.tar.gz
breezy-weather-${{ env.VERSION_TAG }}-gplay.apk
mapping-${{ env.VERSION_TAG }}-gplay.tar.gz
breezy-weather-${{ env.VERSION_TAG }}-fdroid.apk
mapping-${{ env.VERSION_TAG }}-fdroid.tar.gz
```

You're looking for `breezy-weather-${{ env.VERSION_TAG }}-standard.apk` (`-standard` was added as suffix).

**Comment by IzzySoft** [comment]
What I've currently set is:

```yaml
ApkIgnore: /(x86|arm64|armeabi)/i
```

Right, that would no longer suffice, as it would randomly pick 1 and 6-10. So I'd need to add

```yaml
ApkMatch: /standard\.apk/i
```

which then would rule out 6-10, leaving only 1. Does that sound right? Would probably be less complex were the ABI the last element. Then only the `ApkMatch` would be needed :wink:

**Comment by Atemu** [comment]
In the [5.1.0 release notes](https://github.com/breezy-weather/breezy-weather/releases/tag/v5.1.0-beta) it says:

> F-Droid: this version removes support for any weather sources which are not libre and self-hostable, so you are sure to get a 100% FOSS experience and not connect by mistake to a non-free network. There are plans to add this version to the default F-Droid repository.

This is appreciated but one step too far.

Unfree SDKs and other unfree deps are not an option for an F-droid release of course as you have correctly identified.  
Unfree network services however are permitted as long as the client-side code is fully FOSS and as long as "free" alternatives can be chosen:

> This Anti-Feature is applied to apps that promote or depend entirely on a Non-Free network service, or any service which is impossible, or not easy to replace. Replacement requires changes to the app or service. **This antifeature would not apply, if there is a simple configuration option that allows pointing the app to a running instance of an alternative, publicly available, self-hostable, free software server solution.**

https://f-droid.org/en/docs/Anti-Features/#NonFreeNet (emphasis mine)

AFAIUI, your app would not be marked with this anti-feature because OpenMeteo and OpenWeather are available.

Even if it was marked with an anti-feature though, that wouldn't be the end of the world; Geometric Weather was also in F-Droid and marked as NonFreeNet. The purpose of this label is purely informational and does not prevent an app from being included in F-Droid.

**Comment by papjul** [comment]
Yes, I know, I will just quote myself:

> I will just end up making a build variant with no non-free networks (all weather sources except Open-Meteo). I think it’s a good compromise, as:
> 
>     * Weather sources are the elements that can break the more easily, and you know I don’t like how much you have to wait in F-Droid main repository to get updates. It tarnishes the reputation of the app by making users think we can’t fix critical issues fast.
> 
>     * Open-Meteo is a rather well-tested and reliable weather source, so external factors breaking the app are unlikely
> 
>     * This version will only connect to free networks which make it even more FOSS (even though non-free networks are not forbidden and you don’t always tag such apps as “non-free networks”)

Following that decision, I receive a few messages of F-Droid users happy to finally have a flavor of Breezy Weather with only free networks :) And given the philosophy of the main F-Droid repo, I can see how this flavor is totally fitting. But don’t worry, the standard version won't be removed from the Izzy repo which has more lax rules, so both flavors will be available for users to choose, and you can easily switch since it's signed with the same key. I just don't know yet where I will put the version with Google Play Services.

**Comment by EmberHeartshine** [comment]
> I just don't know yet where I will put the version with Google Play Services.

Forgive my confusion, but I would imagine a version of the app with Play Services components would go on the Play Store, no?

**Comment by papjul** [comment]
At the moment, there are no plans to put the app on the Google Play Store. It costs money and we don’t comply (see #31)

**Comment by IzzySoft** [comment]
@papjul thanks for the notification (and apologies for my delayed answer)! So I've set my YAML here to pick from `https://github.com/breezy-weather/breezy-weather/blob/main/fastlane/metadata/android` again.

Please keep in mind that `full_description.txt` is limited to 4,000 characters (yours for `en-US` currently is at 4,133) and will be truncated exactly there. The updater in my repo will "compress" it (removing all line breaks, making it a one-liner HTML), but that's still 4,072 chars.

My configuration here still has

```yaml
  ApkMatch: /standard\.apk/i
  ApkIgnore: /(x86|arm64|armeabi)/i
```

so it would pick `breezy-weather-v5.1.5-standard.apk` currently. Let me know if file names will change, or if I should pick a different one.

> I don't know how exactly you will set it up to make reproducible builds

I don't know who told you about reproducible builds in my repo, but thanks for the hints :stuck_out_tongue_winking_eye:

**Comment by IzzySoft** [comment]
Btw: Do you consider mentioning the IzzyOnDroid repo with your [INSTALL](https://github.com/breezy-weather/breezy-weather/blob/main/INSTALL.md) documentation? It carries the "Standard" flavor as you know. "Delay for updates" would be max 24h. Your app [is here](https://apt.izzysoft.de/packages/org.breezyweather/).

**Comment by papjul** [comment]
> Please keep in mind that `full_description.txt` is limited to 4,000 characters (yours for `en-US` currently is at 4,133) and will be truncated exactly there. The updater in my repo will "compress" it (removing all line breaks, making it a one-liner HTML), but that's still 4,072 chars.

I will check!

> My configuration here still has
> 
> ```yaml
>   ApkMatch: /standard\.apk/i
>   ApkIgnore: /(x86|arm64|armeabi)/i
> ```
> 
> so it would pick `breezy-weather-v5.1.5-standard.apk` currently. Let me know if file names will change, or if I should pick a different one.

It's correct!

> > I don't know how exactly you will set it up to make reproducible builds
> 
> I don't know who told you about reproducible builds in my repo, but thanks for the hints 😜

Multiple discussions here, I was talking about the landing in F-Droid default repo. I moved that discussion to the RFP.

> Btw: Do you consider mentioning the IzzyOnDroid repo with your [INSTALL](https://github.com/breezy-weather/breezy-weather/blob/main/INSTALL.md) documentation? It carries the "Standard" flavor as you know. "Delay for updates" would be max 24h. Your app [is here](https://apt.izzysoft.de/packages/org.breezyweather/).

It used to. However, following a discussion in #796, it seems that having Breezy Weather in too many repos and an F-Droid app configured to auto-update will lead to switch between flavors.

So, to avoid this, I'm heading towards having the app only in the default F-Droid repo + user can add dedicated repo to the flavor they want for faster updates. That way, an user who has your repo configured but only wants the F-Droid flavor of Breezy Weather won't auto-update to the Standard flavor from your repo.

The other way would be to make a different `applicationId`/signature for each flavor, but I like the ability to easily try multiple flavors, especially as we don't have an import/export feature.

What do you think?

**Comment by TheLastProject** [comment]
I trust the IzzyOnDroid repo much more than the f-droid.org repo. I don't want to add another third party repo just to get updates and I don't want to be forced to switch to the f-droid.org repo, I want to keep using IzzyOnDroid as it has a good reputation for security and fast updates.

I strongly recommend using different package names for different flavors (look for example at Aves Gallery, they use a different package name for their F-Droid flavor too). It even makes trying multiple flavors easier as you can install them at the same time due to different package names (which I do with my own app while developing, I have a .debug package name, so I can run both production and debug and compare easily).

Using the same package name for multiple flavors is asking for problems (for example, how would you deal with an IzzyOnDroid user getting an F-Droid update when they use a weather provider not included in the F-Droid build? Lots of complex edge cases that you could never truly automatically fix without upsetting at least some user).

**Comment by licaon-kter** [comment]
> I trust the IzzyOnDroid repo much more than the f-droid.org repo

:no_mouth: 

> how would you deal with an IzzyOnDroid user getting an F-Droid update when they use a weather provider not included in the F-Droid build?

the repo overhaul in F-Droid Client will allow users to select to only get updates from one particular repo per app

**Comment by licaon-kter** [comment]
you can test Debug builds from the nightlies: https://gitlab.com/fdroid/fdroidclient-nightly

**Comment by TheLastProject** [comment]
As a Breezy Weather user who uses both IzzyOnDroid and F-Droid (but prefers IzzyOnDroid), I want to have the full Breezy Weather experience without ever having to worry about randomly being switched to the more limited F-Droid flavor. That's my use case :)

I know it is currently unlikely, but if F-Droid ends up improving their build speed, the fact that both these flavors have the same `applicationId` means an F-Droid client could randomly switch me from the IzzyOnDroid to F-Droid version. This would cause me to lose functionality (either silently or I guess with a notification from Breezy Weather) and force me to take manual steps (going inside the F-Droid client to switch back to the IzzyOnDroid version) to rectify this issue. Even more likely, F-Droid users would be switched to the IzzyOnDroid version all the time even if they maybe wouldn't want this if they just have the repo enabled (which many people do).

Licaon_Kter's promised feature of an option to limit an app to a single update source doesn't truly fix this as it is a manual step you would need to do after realizing you somehow switched version. Because being randomly switched flavor is something that's not expected behaviour (users generally expect to not have their app source randomly change) it means many users may not realize this is what they need to do. It also may not be supported by other clients, which many people use.

I disagree using different `applicationId`s is more confusing to the user, as it prevents the issue that happened in #796. Using different `applicationId` would mean that the user choosing to get the standard version from IzzyOnDroid will never be switched to the F-Droid version and vice-versa without manual steps. It is to me way more unlikely that an user would be confused by multiple versions being available than by randomly gaining and losing functionality based on whatever source happens to provide the update first (which I think is shown very well in #796: an user being surprised by being swapped to another flavor).

The only "risk" of confusion is that users could see multiple entries for Breezy Weather in their client. However this can be worked around easily by setting a different metadata name in the Fastlane, so the F-Droid flavor can be called something like "Breezy Weather (F-Droid)" in clients (you could also have F-Droid specific descriptions in the client based on fastlane for the feature difference). Using multiple `applicationId`s is, as I can see it, the only way to make it easy for users to choose between standard and F-Droid and not be swapped unexpectedly.

**Comment by licaon-kter** [comment]
@TheLastProject you can read my comment again pls, I guess we are lost in translation.

Or straight test the Debug version as linked

F-Droid would love feedback on this functionality, specially from repo users

**Comment by papjul** [comment]
Is there some document that explains why the build process of the main F-Droid repo is so slow? There are tons of forum discussions but I don't easily find a clear explanation, why and what steps are being taken to remediate it. Or if not possible, why. So we know if something can/will change in the future.

My arguments against different `applicationId` are still valid. Following the issue you mentioned, the app will be removed from Izzy repo, so there is no risk an user of the `fdroid` flavor will be switched to the standard version by mistake, since it will no longer be in that repo. The only way an user would be switched to a different flavor is:
- if build process of the main repo is very fast (currently I find that it's about a week between publishing on GitHub/GitLab/etc to the actual push of the release in the F-Droid main repo, but it may be even more if there are build issues)
- if the user has installed one of our repos (for a specific flavor)

> Licaon_Kter's promised feature of an option to limit an app to a single update source doesn't truly fix this as it is a manual step you would need to do after realizing you somehow switched version.

Once the feature is available, I can add instructions to the README of our F-Droid repos, so that no one has to be find out they have been switched to the wrong (although it seems very unlikely given the current delay for new updates).

**Comment by licaon-kter** [comment]
The F-Droid workflow is here: https://gitlab.com/fdroid/wiki/-/wikis/FAQ

By now I find the "speed" argument a moot point as it appears to be used as a last _trump card_ in any discussion.

Yes, developer repo, Izzy's, are faster. That does not entail anything regarding app availability in _other_ repos.

**Comment by TheLastProject** [comment]
Removing the app from IzzyOnDroid really sucks... Removing the app from such a popular repo that provides fast updates is a huge loss to the whole community and leaves existing users without an upgrade path... That's not a "solution" to me, that's a loss...

**Comment by IzzySoft** [comment]
> the app will be removed from Izzy repo

Only in two cases: 1) it becomes too large for my repo, or 2) you request I should remove it. Especially if it's two different build flavors, I tend to keep the app in my repo even when it became available at F-Droid (see e.g. Aves, which is a comparable case).

I fully agree with Sylvia here that removing it from my repo would be a loss. Folks would be forced to either use the limited version from F-Droid, or to add yet another repo just for a single app – something I've often heard they dislike.

> By now I find the "speed" argument a moot point

Have you looked at the build monitor recently?

**Comment by papjul** [comment]
> By now I find the "speed" argument a moot point as it appears to be used as a last _trump card_ in any discussion.

I don't agree at all. This is a real frustration point, both for developers and users.

From the developer point of view, it tarnishes the reputation of our app as being unable to deliver fixes to critical issues fast. And what would happen if an app has a vulnerability and we had to wait for a week to get it to our users? I know that Microsoft does that, but I don't think they are a good example.

From the user point of view, you already quoted them yourselves, they keep complaining about the delays, especially when there are bugs. Because most of us work for free on our apps on our limited free time, we can't make intensive testing (we should, but you know, life).

And even in real life, when I discuss with friends who also use F-Droid, we really like what F-Droid does for the open source, but the delay of updates always come in the conversation. This is really not trivial.

I went through the FAQ quickly, should I understand that limited resources for building + long workflow makes it almost impossible to reduce the delay?

An idea I have is that for reproducible builds, you could assume that if you reproduced once, following versions will also reproduce and you push them immediately (maybe do some basic scanning first) and build afterwards to confirm. Maybe on a second official F-Droid repo in case not everyone agrees with that.

> Have you looked at the build monitor recently?

I made my measurements about two months ago, has something changed since then?

> Removing the app from such a popular repo that provides fast updates is a huge loss to the whole community and leaves existing users without an upgrade path... 

The upgrade path is our repos. They are even faster, since updates are immediately published in the next few minutes the release is made.

> Only in two cases: 1) it becomes too large for my repo, or 2) you request I should remove it. Especially if it's two different build flavors, I tend to keep the app in my repo even when it became available at F-Droid (see e.g. Aves, which is a comparable case).
> 
> I fully agree with Sylvia here that removing it from my repo would be a loss. Folks would be forced to either use the limited version from F-Droid, or to add yet another repo just for a single app – something I've often heard they dislike.

That was my plan to keep it on your repo, but what if someone has your repo installed but only wants the flavor that has no risk to connect to non-free sources? I know a few users who want the `fdroid` flavor only (I don't know if they have your repo set up though).

**Comment by Atemu** [comment]
> From the developer point of view, it tarnishes the reputation of our app as being unable to deliver fixes to critical issues fast. And what would happen if an app has a vulnerability and we had to wait for a week to get it to our users? I know that Microsoft does that, but I don't think they are a good example.

This is an age old discussion that's been discussed many o' times but in the end, dedicated software distributions such as F-droid provide value through their processes and not having them is simply not an option. We can have it again but the result won't be any different.

Here's a recent example of what happens when poorly run distros prioritise speed over quality: https://www.gamingonlinux.com/2024/03/knock-knock-whos-there-more-scam-apps-on-canonicals-snap-store/

I fully understand the frustration of updates not being quite as quick to roll out and users reporting bugs that have been fixed weeks ago. There isn't really a way around it though.  
If you're *really* worried about your reputation being tarnished, you can nicely ask distros to publish your app under a different name if you must. (Or force them via a trademark but please try asking nicely first.)

I'm a distro maintainer myself (GNU userspace, not Android) and, while I don't know how the processes work for F-droid, I can tell you that collaboration with the distros will usually help everyone involved; including the speed at which users receive updates.

If there was an update to a package for example and the upstream maintainer of the packaged application came in and verified that it works as expected, that would give me as a distro maintainer huge confidence to apply that update. I don't know the bottlenecks for F-droid as I'm not involved with them but in our case, this would speed up processes by a *lot*.

> An idea I have is that for reproducible builds, you could assume that if you reproduced once, following versions will also reproduce and you push them immediately (maybe do some basic scanning first) and build afterwards to confirm.

This is not a sane policy for a number of reasons. Sorry.

It's like asking a scientific journal to not peer-review a paper because the author has passed peer-review before.

Distros would love to fully trust upstream maintainers but independent scrutiny is a fundamental pillar of software distributions.

**Comment by papjul** [comment]
Thanks for your inputs, that changed my mind a bit!

I don't question the long integration process that ensures that safe apps are integrated.
But delayed updates don't prevent an included app to become later a scam or integrate non-free blobs or failing to comply with an inclusion criteria. This happened in the past and it is inevitable. Geometric Weather was in this situation for years without anyone noticing. But sure, it's probably a bit more safer with the scanning tools and other stuff that other distributions don't have necessarily.

It's just that from my side, apps like mine can break by other external factors.
Other such similar apps include:
- NewPipe, which has their own repo like mine
- Tachiyomi, which made their release on F-Droid, however sources more likely to break and have frequent updates were made as extension APK (not released on F-Droid). This is an idea I still haven't fully investigate for the future. This looks like an immense task. At least in the current state where I'm doing most of the contribution to the project.

Reproducible builds are at least a good thing that allows user to easily update before they report an issue :)

**Comment by IzzySoft** [comment]
> > Have you looked at the build monitor recently?

> I made my measurements about two months ago, has something changed since then?

That comment was directed at @licaon-kter – and yes, it has. Not for the better, though the current situation hopefully is resolved soon (reason is known I'd say). But such things happen, and cause further delays. I'm not saying that to blame anyone, just stating that "3..5 days" is the optimistic case. With my repo, 24h is the pessimistic case (which happens when you push a new release just after my updater passed by).

One more word on why it would be a good idea to keep the app in my repo, as that just popped up in an unrelated issue elsewhere. Another reason why (some) people prefer using my repo for multiple apps versus single-app repos is also the added scans and security checks here (see [What about security?](https://apt.izzysoft.de/fdroid/index/info#security) in my repo's details page). Such independent scans (which your repo cannot provide, at least not _independently_) give them added confidence. The literal quote if you want it: "I wouldn't use their repos because they don't have your checks."

> an included app to become later a scam or integrate non-free blobs or failing to comply with an inclusion criteria.

Which is part of what those scans are for. Many a dev was thankful for me reporting such things then, as those blobs came in unintended (by some dependency dragging in other dependencies). In some cases even F-Droid was thankful, as my scanners cover things F-Droid's don't (and vice-versa, as mine do not scan the source code).

**Comment by Atemu** [comment]
> delayed updates don't prevent an included app to become later a scam or integrate non-free blobs or failing to comply with an inclusion criteria.

Note that delayed updates are mostly a means to an end, not the end themselves.

The actual reason for the delay is the time the important internal processes of a distro require such as:

- independent review
- additional QA
- integration with the rest of the ecosystem (less of a concern in the context of Android)
- independent trusted build

Of course there can sometimes be other reasons such as a package maintainer simply having a bad feeling about a major update where they'd rather wait for the .1 patch to arrive to avoid delivering bugs to users and sometimes it's also just simply a matter of whether someone is available to press the button.

(There is also the concept of stable distros where users expect software to *not* update except for critical fixes in order for (software- aswell as user-) interfaces to not break within a certain lifecycle. This isn't really widespread in the Android ecosystem; all Android-App distros I know of are a rolling release.)

Again, I can't speak for F-droid but if you want your software to reach distros' users faster, it's in your best interest to aid distros in speeding up all of these steps. If you had a suite of E2E-UI-Tests which a distro could run to verify their build of your code works as expected, that could theoretically speed up the QA process a good bit for instance.

**Comment by papjul** [comment]
Your arguments are valid... but so are opposite ones. When we add an app through your repo, we have to trust both you and the developer. We move a bit of trust off the developer onto you. Other reasons I heard against are "I only want Breezy Weather, I don't want to add a repo full of other apps, some of which don't fully fulfill F-Droid criteria for inclusion". (I just wanted to show you another point of view; I know that you are not trying to tell me to remove my repos but to offer choice)

I have plans to add scans to the workflow. However, you're right, they are not *independant* meaning I could just remove the workflow and publish manually a scammy update. That wouldn't come unnoticed to someone looking at the commit history, but autoupdate mechanism makes it impossible to warn users (at least in time).

That's why I thought about making our repos externally managed. Someone (for example Julian) would manage the repo without the Breezy Weather team having any right on it, with an automated action with scans and only publish if the build is reproducible (which would be somewhat similar to F-Droid default repo + Izzy scans but with faster updates and the ability to build the standard flavor). However, that's always about the same thing, about who you trust. Some may say "I don't know who this Julian is". So that idea was dropped.

In the end, the main issue about having the app spread over many repos is the auto update mechanism. I need to review with the rest of the team about our options. Maybe we could store the current flavor, and if we detect it changed, we show a popup to let the user confirm or show them instructions to revert that change.

I don't think it could happen very often with the update delay thing, although there is one case where it could happen often if we keep it on Izzy repo: if someone wants the `gplay` flavor from our repo, they may autoupdate to the `standard` one from your repo, given the short update delay of both our repos.

> "3..5 days" is the optimistic case

I often see that ~4 days thing, but just to be clear, it's only the length of the cycle, right? It doesn't represent the actual delay. If you being a new cycle at some point, and the developer pushes an update 1 hour after, they will have to wait for the next cycle, meaning the actual delay is somewhere between 3 and 10 days *if* the build suceeds and there is no ongoing problems, right? That would be on average closer to what I measured (a week). I'm not trying to blame anyone, I know that you're doing your best, but at the same time, I don't want you to think "it's not that bad" (when you say "moot point").

Tl;dr: reviewing internally.

**Comment by IzzySoft** [comment]
> we have to trust both you and the developer

In a way. In case of my repo, as the APKs are taken directly from you, the signature should ensure it's mostly you (the developers having signed the APK) to be trusted, while my repo just offers an additional way of distribution plus additional scans providing some insights. The code used for that [is public](https://gitlab.com/IzzyOnDroid/repo/-/tree/master/lib). In case of F-Droid, the same can be said with reproducible builds established, as then again it would be the APK you've signed that is distributed. Here it gets a little different without reproducible builds, as then F-Droid would sign the APKs and thus you'd also need to trust them they didn't "tamper" with the code.

> Other reasons I heard against are "I only want Breezy Weather, I don't want to add a repo full of other apps, some of which don't fully fulfill F-Droid criteria for inclusion".

I've seen that argument (it was posted on the Fediverse), yes. But as you wrote yourself:

> you are not trying to tell me to remove my repos but to offer choice

Exactly. So both are covered :wink:

> if someone wants the `gplay` flavor from our repo, they may autoupdate to the `standard` one from your repo, given the short update delay of both our repos.

Which, as outlined before, can be easily avoided using dedicated package names. As is done e.g. with Aves (a gallery app being in F-Droid's repo with one flavor, carrying the `.libre` suffix in its package name, and with a different flavor in my repo without that suffix). No risk of accidentally replacing one with the other. You could even install each alongside the other. Having an option to export/import settings, switching to a different flavor would also be easy.

> meaning the actual delay is somewhere between 3 and 10 days _if_ the build suceeds and there is no ongoing problems

That's correct. For my repo it would mean ~24 hours if your update "just missed" my updater's run, or less if it didn't.

**Comment by papjul** [comment]
> You could even install each alongside the other.

If you end up having two flavors installed alongside, it's more likely an error. If it's not, you will still have a hard time distinguish them as they have the same logo and name.

> Having an option to export/import settings, switching to a different flavor would also be easy.

We don't have such an option 1) it needs someone to develop it 2) ensuring it's safe from injections 3) ensuring it's backwards compatible (so we have to think the feature the rightest way from the beginning) 4) we need to handle all the hassle of file permission scope of newer Android SDK. 5) it's probably wise to tie this development to the overhaul of the Settings to avoid having to go back on some choices (which would mean additional development time later on)

It's already a highly requested feature although we have the same signature and applicationId everywhere. So if we ever need to make different applicationId, this feature will become mandatory to not have lots of complains (that I would understand).

I'm not interested in developing this feature, and I have to either develop it or get lots of complains. None of the outcomes are something welcome. Especially as Geometric Weather was pulled off the default F-Droid repo and I understand users don't want to wait months for a development/a decision.

**Comment by IzzySoft** [comment]
> If you end up having two flavors installed alongside, it's more likely an error.

I didn't say you *have to* – just that it *would be possible* if one *wants to*. Even if just to check what the differences are and then decide which one to keep.

> you will still have a hard time distinguish them as they have the same logo and name.

Guess I as a non-dev do not have to explain to you that, given they are built from different flavors, this is something easily solved in those flavors :wink:

> 4) we need to handle all the hassle of file permission scope of newer Android SDK.

Huh? For exporting a single JSON file with the settings inside? It's not that this means maintaining a full directory tree. It's a single request to pick a directory (using the file picker), which the app then gets granted read/write access to. So you can store the export file there, and can read it back from there. I'm sure there are thousands of apps already using it that way.

That JSON could also help you solve "3) ensuring it's backwards compatible" with import simply ignoring what it doesn't know about. So you'd only have to take care it's "forward compatible" (missing options using defaults, or simply keep what is already set). Which maybe leaves instances of "renamed properties" – but those could simply be "kept constant" in the JSON.

> 5) it's probably wise to tie this development to the overhaul of the Settings

That makes sense, yes.

> So if we ever need to make different applicationId, this feature will become mandatory

Again, makes sense, yes. And let me stress a last time, that unique `applicationId`s for each flavor would solve the issue of "accidental cross updates". Don't forget that PlayStore tries those as well, only checking whether the `applicationId` matches – so people having both active, run the risk of having PlayStore replace their favorite flavor without even asking them. Happened quite a lot recently with multiple apps – and sometimes only failed because the signature didn't match (PlayStore was even too stupid to check *that* in advance and still tried to replace).

**Comment by papjul** [comment]
So we came up with some kind of mix between the two solutions, using the `versionCode`.

Starting from V5.1.8:

Old system: `50108`

New system:

| Build type | Flavor             | VersionCode |
|------------|--------------------|-------------|
| Debug      | Any                | `501080`    |
| Release    | `fdroid`           | `501081`    |
| Release    | `basic` (standard) | `501082`    |
| Release    | `gplay`            | `501083`    |

It’s not possible to downgrade to a lower VersionCode. So this means that you will no longer downgrade to a less complete version if you have multiple repos.

One case is not covered: if you have Izzy repo and you only want the `fdroid` flavor from default F-Droid repo. But that shouldn’t that big of a deal for the very few affected people, since the standard version can be used with only free-net sources.

Bonus point of the same `applicationId`: ensures an user who removed Izzy repo can still receive updates from F-Droid default repo to not remain with a potentially insecure version.

Let me know if that sounds good to you and I will make a new release. I know no solution is perfect, but at least, we can still change that later, as compared to deciding right away to make different `applicationId` (merging back to one `applicationId` would not be possible).

**Comment by papjul** [comment]
Breezy Weather is not on the Play Store, and even if it was, that wouldn't be a problem updating to the `gplay` version, since it only contains a Google Play Services component the user already has on their device.

You make it look like it's an easy task and there are less troubles than having a single `applicationId`, but that's the complete opposite, as already explained in length:
- Different `applicationId` allow to have 3 different apps of Breezy Weather, which is an error, uninstalling the other 2 is the only logical thing to do. Even if it was logical to have 3 different apps of Breezy Weather, they would have the same icon and app name, which would not allow to easily distinguish them. And I won't make different icons and app names for no actual real-life use case.
- It needs someone to write an Export/Import feature, which would probably not be hard for the preference part (key/value), but as for the location part, it's a lot of headache as we need to ensure the consistency of the data imported and I have no actual solution in mind for quite a few inconsistencies. I prefer to spend 15 minutes adding a different versionCode than spending months making an approximately reliable way to make it possible to use Breezy Weather in a way that is not intended with 3 different apps. I'm working on my free time on this project and I want to remain free to develop what I want to. I don't want someone to tell me or urge me to "just make an export/import functionality, done". I have no intention to work on this in priority, so if it takes 1 year, 2 years or more for this issue to come at the top of my priority list, that's my choice.
- We won't agree on this one, but if someone removes your repo, they will no longer receive updates with the security issues that can result from it. If someone really needs the `standard` version after removing Izzy repo, they have the option of our repo. And if they want neither of them, well, that's not something different `applicationId` can solve anyway.

With the versionCode system I propose, you only update to a more complete flavor, so there is no risk of having your app replaced with a less complete flavor, so I don't see the issue?

> different package names (`applicationId`s) would save you from all this trouble

So, which troubles remain to be solved?
In the opposite, I listed quite a few against different `applicationId`.

**Comment by obfusk** [comment]
> With the versionCode system I propose, you only update to a more complete flavor, so there is no risk of having your app replaced with a less complete flavor, so I don't see the issue?

Only if both have the exact same underlying version. You can still get a less complete flavour if its source happens to have a newer version -- since *any flavour* of e.g. 5.1.9 will have a higher version code than *any flavour* of 5.1.8. So if your sources get slightly out of sync wrt the version, you can still get unintended cross-updates.

**Comment by papjul** [comment]
All of our repos update immediately (the only way to get `gplay`), `standard` from Izzy updates at worst every 24 hours and `fdroid` the minimum is 3 days.

**Comment by bblacher** [comment]
> All of our repos update immediately (the only way to get `gplay`), `standard` from Izzy updates at worst every 24 hours and `fdroid` the minimum is 3 days.

I've been following this discussion for a while and it seems that by that logic the versioning system doesn't do anything because even without it you would never get downgraded to a "lesser" version.

As a user I'd just be glad if the app would be added to the official Fdroid repo as that is where I get all of my apps from and I don't like keeping other repos around.

**Comment by obfusk** [comment]
> All of our repos update immediately (the only way to get `gplay`), `standard` from Izzy updates at worst every 24 hours and `fdroid` the minimum is 3 days.

The user's client could easily see indices that are out of sync even if the repositories themselves are in sync. Automatic updates of indices fail. Network errors happen. etc.

**Comment by papjul** [comment]
> you would never get downgraded to a "lesser" version

If `gplay` and `standard` have the same versionCode, Izzy can (and will most of the time) replace the `gplay` version since most F-Droid clients use the most recent release date when two version codes are identical. My proposition solves this issue by making sure the F-Droid client always pick up the `gplay` version.

**Comment by papjul** [comment]
> > All of our repos update immediately (the only way to get `gplay`), `standard` from Izzy updates at worst every 24 hours and `fdroid` the minimum is 3 days.
> 
> The user's client could easily see indices that are out of sync even if the repositories themselves are in sync. Automatic updates of indices fail. Network errors happen. etc.

So, in the worst case, the Gplay repo is unavailable and Izzy is available, so you get your gplay version replaced with standard, what are the consequences? Until the next successful sync, you will use standard flavor with GPS and/or a slightly less powerful Google Play Services network-provided location compared to gplay version. No one would ever notice since Google Play Services still take the lead when GPS is not available.

**Comment by obfusk** [comment]
What if my client doesn't update for 4 days and replaces gplay/standard with f-droid? From personal experience, this will happen regularly, even if you set it to check for updates daily; weekly would make it even more likely. And that's with a reliable network connection. And just one of the ways I can think of for things to get out of sync.

I wish this scheme worked. It would save you a lot of work. I get why you don't want different appids. I just don't think "no risk of having your app replaced with a less complete flavor" is something you can guarantee with this version code scheme as it relies on nothing getting out of sync, which is unfortunately not something you can control or predict.

**Comment by IzzySoft** [comment]
> So, in the worst case, the Gplay repo is unavailable and Izzy is available

The worst case is that GPlay replaces any installed non-gplay version and you've got all those proprietary things in you wanted to avoid in the first place. Not everybody is able to run a Google-free ROM with all Google-ties removed, but still might want to keep proprietary parts as few as possible.

But do as you wish, I won't argue anymore. You've made your decision already. But when folks start complaining about "unexpected" cross-updates then – please remember that those *are* to be expected. And *will* happen. The only way to be sure to avoid that is using different package names.

:man_shrugging:

**Comment by bblacher** [comment]
@IzzySoft the gplay-flavor isn't available on the Google Play store

**Comment by papjul** [comment]
> What if my client doesn't update for 4 days and replaces gplay/standard with f-droid? From personal experience, this will happen regularly, even if you set it to check for updates daily; weekly would make it even more likely. And that's with a reliable network connection. And just one of the ways I can think of for things to get out of sync.
> 
> I wish this scheme worked. It would save you a lot of work. I get why you don't want different appids. I just don't think "no risk of having your app replaced with a less complete flavor" is something you can guarantee with this version code scheme as it relies on nothing getting out of sync, which is unfortunately not something you can control or predict.

So you have to 1) only sync once a week 2) have other repos fail to sync (not something that ever happened to me on one repo but not the other), so that's rather improbable, but even in that case, you will have a feature in the next version of the F-Droid client to select from which repo you want to get updates from. I can't take care of edge cases like that compared to other issues that would affect all users with other solutions :/

**Comment by papjul** [comment]
> > So, in the worst case, the Gplay repo is unavailable and Izzy is available
> 
> The worst case is that GPlay replaces any installed non-gplay version and you've got all those proprietary things in you wanted to avoid in the first place. Not everybody is able to run a Google-free ROM with all Google-ties removed, but still might want to keep proprietary parts as few as possible.

Why would someone add the Gplay repo if they don't want the Gplay version?


> But do as you wish, I won't argue anymore. You've made your decision already. But when folks start complaining about "unexpected" cross-updates then – please remember that those are to be expected. And will happen. The only way to be sure to avoid that is using different package names.

So, can you agree that the versionCode new situation is better than all flavors sharing the same versionCode? It seemed to me you were arguing against and everyone seemed to thumb up. If that's the case, I can revert the commit, I don't care, but it looks to me that you want to avoid cross-updates to happen and my solution solves almost all cases, except a few very rare of occurrences of out-of-sync repos that will be solved with next version of the F-Droid client

**Comment by obfusk** [comment]
> only sync once a week 

I have it set to sync daily; it doesn't.

> have other repos fail to sync (not something that ever happened to me on one repo but not the other)

This has happened to me more than once.

> you will have a feature in the next version of the F-Droid client to select from which repo you want to get updates from

There are other clients than the official one.

> I can't take care of edge cases like that compared to other issues that would affect all users with other solutions :/

It's your choice. I get you're working in your free time. I don't want to make your life more difficult. I just want to make sure you know that these edge cases _will_ affect some of your users, so you -- and your users -- can make an informed choice.

**Comment by papjul** [comment]
> This has happened to me more than once.

With a daily sync, you already had Izzy repo failing to sync for a whole week but not F-Droid official repo? That's the case we are talking about.

> There are other clients than the official one.

If the official client implements this feature, it's because there is a need for it. I understand maintainers of other clients may not have the time or wish to implement it, though. Even without that feature, I'm pretty sure this will almost never happen, and if it does, there is a simple way for the user to get back to the flavor they want.

> It's your choice. I get you're working in your free time. I don't want to make your life more difficult. I just want to make sure you know that these edge cases will affect some of your users, so you -- and your users -- can make an informed choice.

These edge cases would be (almost) inexistant if we sticked to my plan to remove the app from Izzy repo and only keep on default repo, so that users wanting another flavor would make an informed choice with our own repos. But you highly recommended to keep it and I understood the arguments. That's why I came up with the versionCode new system which should make it possible to cover almost all cases. If it doesn't, you will still have to rely on the future feature of F-Droid client. And if you use a client that doesn't have that feature, in almost all cases you will never encounter any issue anyway.

Note that as I said previously, making today one single `applicationId` doesn't prevent us to change our mind later if we find out it causes too much trouble. The other way around (starting with different `applicationId` right away) causes more problems at the moment.

**Comment by IzzySoft** [comment]
> if we sticked to my plan to remove the app from Izzy repo

Oh, you would make a lot of your users happy that way I'm sure :wink: 

![image](https://github.com/breezy-weather/breezy-weather/assets/6781438/c021a929-05b3-45b3-869a-30ba80cdca7e)

(stats from a single mirror of my repo, so there are 2 more servers to consider)

> doesn't prevent us to change our mind later if we find out it causes too much trouble.

Which will most likely happen sooner than you think (I was already asked if I'd take bets :see_no_evil:). I've seen that too often already, sorry. But as I wrote before: your choice. You cannot say nobody told you. Several long-year maintainers did so now.

Note it will be not that easy changing the appId with the F-Droid listing, once your app is in. That will take at least a few weeks, from my experience (though @licaon-kter might feel triggered now to prove me wrong and make it in 10 days :stuck_out_tongue_winking_eye:).

That said, good luck!

**Comment by papjul** [comment]
> (stats from a single mirror of my repo, so there are 2 more servers to consider)

Am I understanding right that these are the stats of all time? So the number of regular users (users who did not uninstall it after trying it) from your repo is closer to 1000 on average? With 3 servers, that would make ~3000? Additionally, a few users said they will remove your repo (or mine) once it lands in F-Droid default repo, so it will probably be even lower. What are the stats on average of a rather popular app on the F-Droid default repo in comparison?

> I've seen that too often already

I'm sure you did and I can understand people would be upset (at least I would) to update to a newer versionCode flavor with trackers or any other unwanted non-removable blob or antifeature. But that's not our case, we only have non-free networks that can be ignored. Do you have a few examples of similar cases?

> > and my solution solves almost all cases
> 
> Does it? Even if your solution worked perfectly, it would only prevent "having your app replaced with a less complete flavor". Obviously, users will be upset if they lose functionality. But users that _explicitly chose to install a "less complete flavour"_ will still be unpleasantly surprised to get cross-updates, even if they don't lose functionality. After all, they _chose the less complete flavour for a reason_, which is the whole point of providing multiple flavours isn't it?

Flavors in Breezy Weather are mostly there to comply with inclusion policies or specificities of repos. Putting the `gplay` version aside since that requires conscious decision of adding the gplay repo, we are only talking about fdroid vs standard. Fdroid flavor was made due to popular demand to publish on F-Droid default repo, as it would help workaround API keys issues; avoid having antifeatures; avoid having delayed hot fixes for proprietary less stable sources. Having only free-network sources is a bonus point for some users, but I don't expect there are a lot of "Richard Stallman"-like users of both default F-Droid AND Izzy repos (which is the case in which it can happen). Even if there are, it doesn't prevent them connecting to only free-net sources, just like you can choose to connect only to free-network websites with the flavor of a web browser that also supports connecting to non-free network websites. So that’s another niche case.

____

Linux distros have the same issues with additional repos default behavior (I can't speak for all of them ofc, but at least for my distro). One repo with a customized library compared to main repo can replace your library from main repo (and you probably won't even notice it). There is no "applicationId" to manage that since they install the files at the same place. Just like Linux users, Android F-Droid users should be aware of that when using additional repos, and should have a way to whitelist/blacklist per repo.

The more I look at it, and the more I think whitelist/blacklist system per repo should have been a core feature from the moment repos were introduced. It hasn't, because there were workarounds for years:
1) using your own signature for F-Droid default repo, but that will be less and less of a workaround with reproducible builds (that are now encouraged)
2) asking maintainers to make different `applicationId`s but I will repeat myself, this is not a normal situation to be able to have 3 different "same app" at the same time

The correct answer to the very few affected people is to use the whitelist/blacklist per repo feature. An alternative is what Obtainium does. You add one app and you decide which source (including F-Droid repos) you want to get updates from. But Obtainium works a bit differently, you can't "discover" apps, you only add apps one by one manually.

I tried the debug version of the F-Droid client, and I'm not sure the feature will work exactly the intended way. I see the term "Preferred repo", does it mean it still fallbacks to other repos if versionCode is higher? If that's the case, I'm not sure to understand the point? I was more expecting in the repo details interface to have a choice between "Include all apps" (by default) or "Only specific apps" and the user choose which ones and all the other apps are hidden in the F-Droid client (just like what you already do with the "Some apps were hidden to respect your antifeatures preferences" message). I'm sure this would be useful to many users of the Izzy repo who are not particularly interested in discovering other apps but only want updates for 1 or 2 specific apps ;) And it also prevents a repo suddenly getting updates from an app it didn't have initially when the user added the repo (this is what happened to us with Julian repo).

**Comment by IzzySoft** [comment]
> Am I understanding right that these are the stats of all time?

No. Just for the last ~3 weeks (since the mirror was established on March 4).

> What are the stats on average of a rather popular app on the F-Droid default repo in comparison?

I cannot tell. I'm currently [not active at F-Droid](https://gitlab.com/fdroid/admin/-/issues/447), and I never had access to the logs or stats there.

> we only have non-free networks that can be ignored.

So they are all disabled by default? And no Google connections established accidentally by the GMS components (and you're sure about that, as you cannot check the GMS source code)?

> Do you have a few examples of similar cases?

I didn't keep records, so no, sorry.

> I don't expect there are a lot of "Richard Stallman"-like users

Ah, nice way to put it :rofl: But still, there might be some of those. I'd be quite upset finding the `gplay` flavor on one of my devices, for example. Or if I'd suddenly find sources enabled I wanted to avoid when choosing the other flavor.

> Linux distros

are a little different. If I'd knew I might run into such a case, I could avoid that using apt pinning (guess something comparable exists outside the Debian world as well).

> There is no "applicationId" to manage that

It's called "package names" :stuck_out_tongue_winking_eye: But as just outlined, to manage our case-at-hand it would be apt pinning, or its corresponding counterpart (with `yum` e.g. you can do it in the config of the corresponding repo).

> this is not a normal situation to be able to have 3 different "same app" at the same time

It's rarely needed, but I'd not call it abnormal. One of my favorite apps does it as well, btw (the gallery app Aves by Thibault Decker).

> The correct answer to the very few affected people is to use the whitelist/blacklist per repo feature.

Which can be found where exactly? I've never seen that in any client.

> "Preferred repo", does it mean it still fallbacks to other repos if versionCode is higher?

It's not yet fully in place, so I can only tell you how it's planned (according to what I remember): If you install an app the first time and do not specify which repo to take it from, it will be installed from the "preferred repo". Once an app is installed, it will be remembered what repo it was installed from, and updates will be bound to that until you explicitly switch.

Again, this a) is not available yet, and b) not with other [F-Droid client](https://android.izzysoft.de/applists/category/named/apps_markets#group_1181)s (at least not to my knowledge).

**Comment by papjul** [comment]
> So they are all disabled by default?

If you set up your location list with the `fdroid` flavor, you won't have any connection to non-free net sources. In the event you add a new location from the standard version, you will always be presented with a dialog letting you choose which sources you want to connect to, before any connection are made.

> And no Google connections established accidentally by the GMS components (and you're sure about that, as you cannot check the GMS source code)?

That's a different matter. If you have Google Play Services, your device will definitely make calls to Google Play Services, Breezy Weather being installed or not.

But I think there is some misunderstanding about how Google Play Services and location services work. When you use current location feature (and only in that case), my app will request location to Android system. Android system will provide network+GPS location. Starting from Android 12 (I think), there's a new permission to choose between network+GPS or network-only, but no option to choose GPS-only.

Network-powered location is made by whatever is installed on your system. Google Play Services, Huawei Services, microG Services, Petrouchka Services, etc.

So, if you use current location feature AND you have Google Play Services installed, whatever flavor you use, you will (almost) definitely make calls to Google Play Services¹ at Android system level. Same with Huawei, etc. Only a few GrapheneOS report they can't get a fix because they have a way to disable Google Play Services and no one is able to get a location in 10 seconds through GPS only, but users who choose to get rid of Google Play Services usually know that well.

¹ Calls might be delayed, it's possible Google downloaded an offline database earlier, as I don't know how it works internally, but at some point your device will definitely make calls to Google Play Services if you have Google Play Services installed (logic)

As you can guess, it's not something specific to Breezy Weather, all apps that use the location services work that way... because that's how the native Android location services work.

> I'd be quite upset finding the gplay flavor on one of my devices, for example.

If you have the `gplay` flavor installed, it's because you chose to. If you forgot you did, the instructions are 1) go to repo management of your F-Droid client 2) Remove Breezy Weather Gplay repo

> Or if I'd suddenly find sources enabled I wanted to avoid when choosing the other flavor.

In the very rare event it happens:
1) You can still choose to not use them.
2) If that still bothers you, you can set up your F-Droid client to sync once a day OR avoid having mix up of auto updates and manual repo sync
3) you may want to ask your favorite F-Droid client maintainers how you can choose which repos you can get updates from. If there aren't, wait for it to be implemented
4) or switch to a client/use a complementary client that has a better repo management. Like Obtainium I talked about earlier
5) or ignore that error as it is supposed to be almost impossible to come up again given you need to have a sync issue of Izzy but not the F-Droid official repo AND have auto sync disabled or once a week AND there is a new update of Breezy Weather. Lots of conditions.

If no solution is satisfactory, I'm fine with ignoring these edge cases. I can't solve repo management issues of apps I don't maintain.

> If I'd knew I might run into such a case, I could avoid that using apt pinning (guess something comparable exists outside the Debian world as well).

On my system, I would mask all packages and unmask the packages I want to use specifically from that repo. See? There is always a way, but not in F-Droid which is missing an important feature IMO (that will hopefully be solved soon).

> It's rarely needed, but I'd not call it abnormal. One of my favorite apps does it as well, btw (the gallery app Aves by Thibault Decker).

It's the snake eating its tail. If you provide no way to manage that from the client, maintainers will find dirty hacks like "different signature" or "different applicationId", but if you always do that, you never solve the issue at its origin.

> Once an app is installed, it will be remembered what repo it was installed from, and updates will be bound to that until you explicitly switch.

Perfect! So there will be a way :) Users have to wait a bit, use the debug version or use Obtainium in the meantime. Although I'm not as certain as you that there will be that many users affected (if any at all), given the versionCode management we have, the delay of F-Droid default repo updates, and especially given the stats you showed me.

> not with other F-Droid clients (at least not to my knowledge)

At least, it opens a path for other apps to implement it. You can understand that I can't take on my own missing features of all (F-Droid or not) clients in the world.

If you use a 3rd-party apt client that doesn't have support for pinning and you absolutely need it, what will you do? You will open an issue upstream, and in the meantime either 1) switch to the official client 2) or postpone the installation of that package you wanted

**Comment by IzzySoft** [comment]
> But I think there is some misunderstanding about how Google Play Services and location services work.

That's not the point. I'm aware of the fused location provider. But with GMS being proprietary and its code not open, you cannot give guarantees what else it might be doing. You have to trust what Google says about it. Which I don't. For example, they once said not to collect location data when you switched the corresponding toggle off – and were found with their hands in the cookie jar still collecting location data but simply not showing them.

> If you have the gplay flavor installed, it's because you chose to.

As Breezy is not on PlayStore, I agree here. Leaves the other two flavors to get cross-updated.

> If that still bothers you, you can set up your F-Droid client to sync once a day

Which is the default at least with the official client. And doesn't solve that the indices of your repos and mine will usually have updates much earlier than F-Droid, as their build cycle takes much longer.

> you may want to ask your favorite F-Droid client maintainers how you can choose which repos you can get updates from.

LOL. Maintainers cannot solve that, sorry. That's a question of either packageNames or the client being able to "pin apps to a repo". Which the official client might do soon, but you cannot tell about the other clients. Telling people which client they have to use to not get flavors of your app mingled is not a decent way of solving that.

> given you need to have a sync issue of Izzy but not the F-Droid official repo

From my experience, rather the other way around (see "build cycle" above). So the `fdroid` flavor would be delayed.

> Perfect! So there will be a way :)

Only if you use that specific client. As said, fixing people to a specific client is not what I'd consider decent to do.

> or use Obtainium

and loose all the additional checks performed in the F-Droid repos. Not an option for many.

But we're going in circles here, so I'll stop arguing now (i.e. not answering arguments anymore is not meant as disrespect to you, forgive me, but this is stealing my time and yours as obviously, as stated above, your decision is already fixed anyway. Thanks for at least considering to think about it – but our opinions on this differ too much I guess.

**Comment by papjul** [comment]
> That's not the point. I'm aware of the fused location provider. But with GMS being proprietary and its code not open, you cannot give guarantees what else it might be doing. You have to trust what Google says about it. Which I don't. For example, they once said not to collect location data when you switched the corresponding toggle off – and were found with their hands in the cookie jar still collecting location data but simply not showing them.

I'm not talking about Fused location, and I agree with what you're saying, but this has nothing to do with Breezy Weather. We ask the location services to provide us with their data. If the Android system decides to call Huawei services or Google Play Services, this is nothing we can control. I have none on my device, so I know no such services are called on my device, so it's a device-dependent issue that Breezy Weather cannot solve.

> Which the official client might do soon, but you cannot tell about the other clients. Telling people which client they have to use to not get flavors of your app mingled is not a decent way of solving that.

In other words, if I support F-Droid, then I need to support all its quirks? I'm fine with removing support for F-Droid store to avoid being accountable for that, but I'm pretty sure that's not what users think, especially given -once again- that most of them will not be affected by it.

> fixing people to a specific client is not what I'd consider decent to do.

I'm only suggesting it to users who 1) want both F-Droid and Izzy repos 2) but only get updates from F-Droid repo for Breezy Weather, to use the official F-Droid client or Obtainium. This kind of users might not even exist given the stats you shared.

Following your logic, Google Play users, Amazon users or whatever-store users might also be upset to have to install a F-Droid client just to receive updates for this app.

What would be not decent is dropping the idea of publishing to F-Droid default repo to avoid upsetting some users that might not even exist.

Taking decisions in a project that will upset a minority of people but brings more benefits to many others happen all the time. The opposite is also true: not doing it will upset these many other people that are expecting it.

We have workarounds available. If they are too constraining for them, then it's fine if they choose to not keep the app for that reason.

>> or use Obtainium
> 
> and loose all the additional checks performed in the F-Droid repos. Not an option for many.

Obtainium allows you to use any F-Droid repo. For getting updates from your repo, just input the following URL as your source and you're done:
https://apt.izzysoft.de/fdroid/index/apk/org.breezyweather

> But we're going in circles here, so I'll stop arguing now (i.e. not answering arguments anymore is not meant as disrespect to you, forgive me, but this is stealing my time and yours as obviously, as stated above, your decision is already fixed anyway. Thanks for at least considering to think about it – but our opinions on this differ too much I guess.

Thanks for your time. Yes, the more I looked into it, the more different `applicationId` were off the table, so if it is the only suggestion, then it's a no. I'm open to other suggestions that allow switching between flavors without uninstalling/reinstalling. For example, I suggested different F-Droid clients and versionCode management to handle that better, but there might be other suggestions I didn't think of.

Can we at least agree versionCode by flavor is better than same versionCode for all flavors? Or should I revert?

**Comment by papjul** [comment]
App is scheduled to be available in next cycle of F-Droid default repo.
